persistence/mongo/src/main/scala/za/co/absa/spline/persistence/mongo/dao/LineageDAOv5.scala

```diff
@@ -0,0 +1,217 @@
+            addAll(opTransformations.asJava)
+          })
+        }
+    })
+    lineage
+  }
+
+  private def isProjectOperation(op: DBObject): Boolean = {
+    val hintStrategy = BSONSalatContext.ctx.typeHintStrategy
+    val opClassName = hintStrategy.decode(op.get(hintStrategy.typeHint))
+    opClassName.endsWith("op.Projection")
+  }
+
+  private def getOperationId(op: DBObject) =
+    op.get(Field.mainProps).asInstanceOf[DBObject].get(Field.id).asInstanceOf[UUID]
+
+  val binaryTypeHintStrategy = BinaryTypeHintStrategy(TypeHintFrequency.Always)
+
+  object Field {
+    val child = "child"
+    val children = "children"
+
+    val condition = "condition"
+    val aggregations = "aggregations"
+    val groupings = "groupings"
+    val transformations = "transformations"
+    val orders = "orders"
+    val expression = "expression"
+
+    val t = "_t"
+    val typeHint = "_typeHint"
+    val id = "_id"
+    val opId = "_opId"
+    val sparkVer = "sparkVer"
+
+    val mainProps = "mainProps"
+    val datasetId = "datasetId"
+    val dataType = "dataType"
+    val dataTypeId = "dataTypeId"
+    val elementDataType = "elementDataType"
+    val elementDataTypeId = "elementDataTypeId"
+    val fields = "fields"
+    val text = "text"
+    val value = "value"
+    val name = "name"
+    val exprType = "exprType"
+  }
+
+  object SubComponentV5 {
+
+    sealed trait SubComponentV5 extends SubComponent
+
+    case object Transformation extends Component("transformations") with SubComponentV5
+
+    case object DataType extends Component("dataTypes") with SubComponentV5
+
+    val values: Seq[SubComponent] = sealedInstancesOf[SubComponentV5].toSeq
+  }
+
+}
+
+trait MutableLineageUpgraderV5 {
+  this: VersionedLineageDAO =>
+
+  import MutableLineageUpgraderV5._
+
+  override def upgrader = Some(new VersionUpgrader {
+    override def versionFrom: Int = 4
+
+    override def apply[T](data: T)(implicit ec: ExecutionContext): Future[T] = data match {
+      case None | _: UUID | _: Number => Future.successful(data)
+
+      case Some(o) => apply(o).map(Some(_).asInstanceOf[T])
+
+      case iterable: CloseableIterable[_]
+        if iterable.iterator.isEmpty => Future.successful(data)
+
+      case iterable: CloseableIterable[_] =>
+        Future.traverse(iterable.iterator)(apply).
+          map(new CloseableIterable(_, iterable.close()).asInstanceOf[T])
+
+      case _: DescriptorDBObject => Future.successful(data)
+      case lineage: DBObject
+        if (lineage get Field.id).toString startsWith "ln_" =>
+          upgradeLineage(lineage)
+          Future.successful(lineage.asInstanceOf[T])
+    }
+  })
+}
+
+object MutableLineageUpgraderV5 {
+
+  import LineageDAOv5.{ binaryTypeHintStrategy => hints }
+
+  def upgradeLineage(lineage: DBObject): Unit = {
+    for (op <- lineage.get(Component.Operation.name).asInstanceOf[ju.List[DBObject]].asScala) {
+      val opType = getOperationType(op)
```

Comment:
I'd say if you create a `getOperationType()` than there should also be corresponding `setOperationType()` or alike. 
This piece of code does both getting and setting of the type, so the implementation should be consistent in both directions.

Reply:
`getOperationType` extracts simple class name string of operation similarly to `MutableLineageUpgraderV4#getOperationType`.  Do you suggest to extract rest of the code which sets encoded field t into `setOperationType`?

Reply:
at least add a setter method. So if getType(x) == "Read", then to modify the type you call `setType(x, "Read"). 
In `DAOv4` there is `renameExpressionType` for instance.